Python/compile.c

```diff
@@ -7745,6 +7746,120 @@ assemble_jump_offsets(struct assembler *a, struct compiler *c)
     } while (extended_arg_recompile);
 }
 
+
+// Ensure each basicblock is only put onto the stack once.
+#define MAYBE_PUSH(B) do {                          \
+        if ((B)->b_visited == 0) {                  \
+            *(*stack_top)++ = (B);                  \
+            (B)->b_visited = 1;                     \
+        }                                           \
+    } while (0)
+
+static void
+scan_block_for_local(int target, basicblock *b, bool unsafe_to_start,
+                     basicblock ***stack_top)
+{
+    bool unsafe = unsafe_to_start;
+    for (int i = 0; i < b->b_iused; i++) {
+        struct instr *instr = &b->b_instr[i];
+        if (instr->i_oparg != target) {
+            continue;
+        }
+        if (unsafe && instr->i_except != NULL) {
+            MAYBE_PUSH(instr->i_except);
+        }
+        switch (instr->i_opcode) {
+            case LOAD_FAST_CHECK:
+                // if this doesn't raise, then var is defined
+                unsafe = false;
+                break;
+            case LOAD_FAST:
+                if (unsafe) {
+                    instr->i_opcode = LOAD_FAST_CHECK;
+                }
+                unsafe = false;
+                break;
+            case STORE_FAST:
+                unsafe = false;
+                break;
+            case DELETE_FAST:
+                unsafe = true;
+                break;
+        }
+    }
+    if (unsafe) {
+        // unsafe at end of this block,
+        // so unsafe at start of next blocks
+        if (b->b_next && !b->b_nofallthrough) {
+            MAYBE_PUSH(b->b_next);
+        }
+        if (b->b_iused > 0) {
+            struct instr *last = &b->b_instr[b->b_iused-1];
+            if (is_jump(last)) {
+                assert(last->i_target != NULL);
+                MAYBE_PUSH(last->i_target);
+            }
+        }
+    }
+}
+#undef MAYBE_PUSH
+
+static int
+mark_unknown_variables(struct assembler *a, struct compiler *c)
+{
+
+    basicblock **stack = make_cfg_traversal_stack(a->a_entry);
+    if (stack == NULL) {
+        return -1;
+    }
+
+    // Which locals are function parameters?
+    uint64_t param_mask = 0;
+    PyObject *param_names = c->u->u_ste->ste_varnames;
+    for (Py_ssize_t i = 0; i < PyList_GET_SIZE(param_names); i++) {
+        PyObject *param = PyList_GET_ITEM(param_names, i);
+        PyObject *index = PyDict_GetItem(c->u->u_varnames, param);
+        assert(index != NULL);
+        assert(PyLong_CheckExact(index));
+        long l_index = PyLong_AsLong(index);
```

Comment:
Isn't `l_index == i` always?

Reply:
Yep, it looks like it. The assertion passed, so I removed that code.

Looking at symtable.c, It looks like every `VISIT(st, arguments, ...);` is immediately preceded by a `symtable_enter_block(...)`, so parameters should be guaranteed to be at the lowest indices.